examples/src/app/tree-crdt/TreeView.tsx

```diff
@@ -0,0 +1,222 @@
+'use client'
+
+import { useEffect, useMemo, useState } from 'react'
+import { LayoutNode, NodeBox, Point } from './tree_layout'
+import { ROOT_ID, YTree, YTreeNode } from './ytree'
+import { useMap } from '@y-sweet/react'
+import * as Y from 'yjs'
+import CopyLink from '@/components/CopyLink'
+import Paragraph from '@/components/Paragraph'
+import { Link } from '@/components/Link'
+
+function useDragHelper(callback: (origin: LayoutNode, target: LayoutNode) => void) {
+  const [dragOrigin, setDragOrigin] = useState<LayoutNode | null>(null)
+  const [dragTarget, setDragTarget] = useState<LayoutNode | null>(null)
+
+  return {
+    dragOrigin,
+    dragTarget,
+    mouseOver: (node: LayoutNode | null) => {
+      if (dragOrigin !== null) {
+        setDragTarget(node)
+      } else if (node !== null) {
+        setDragOrigin(node)
+      }
+    },
+    mouseUp: () => {
+      if (dragOrigin !== null && dragTarget !== null && dragOrigin !== dragTarget) {
+        callback(dragOrigin, dragTarget)
+      }
+      setDragOrigin(null)
+      setDragTarget(null)
+    },
+  }
+}
+
+function Arrow(props: { origin: Point; target: Point; offset?: number; animate: boolean }) {
+  let length = Math.sqrt(
+    (props.target.x - props.origin.x) ** 2 + (props.target.y - props.origin.y) ** 2,
+  )
+
+  if (length < (props.target.radius || 0)) {
```

Comment:
When would this be true? How could the distance between two nodes be less than the radius of one of the nodes (unless the layout was messed up) or less than 0?

Reply:
If you create enough children of one node, there can be less than the radius between them. Maybe that counts as “layout is messed up”, but it's easy enough to solve for.

<img width="273" alt="image" src="https://github.com/drifting-in-space/y-sweet/assets/46173/477aef70-359d-455e-9dfc-8d7bf2d7579e">

0 is used here because the radius of a point with no radius set is 0. length will never be less than 0, but we still want the expression to evaluate to `false` rather than `undefined`. We use a zero radius point when drawing the arrow to the mouse position instead of locking it on to a circle.

Reply:
Ahh gotcha